xray/sql_context.go

```diff
@@ -0,0 +1,752 @@
+		return nil
+	}
+
+	db, err := sql.Open(driver, dsn)
+	if err != nil {
+		return err
+	}
+	sql.Register(driver+":xray", &driverDriver{
+		Driver:   db.Driver(),
+		baseName: driver,
+	})
+	initializedDrivers[driver] = struct{}{}
+	db.Close()
+	return nil
+}
+
+// SQLContext opens a normalized and traced wrapper around an *sql.DB connection.
+// It uses `sql.Open` internally and shares the same function signature.
+// To ensure passwords are filtered, it is HIGHLY RECOMMENDED that your DSN
+// follows the format: `<schema>://<user>:<password>@<host>:<port>/<database>`
+func SQLContext(driver, dsn string) (*sql.DB, error) {
+	if err := initXRayDriver(driver, dsn); err != nil {
+		return nil, err
+	}
+	return sql.Open(driver+":xray", dsn)
+}
+
+type driverDriver struct {
+	driver.Driver
+	baseName string // the name of the base driver
+}
+
+func (d *driverDriver) Open(dsn string) (driver.Conn, error) {
+	rawConn, err := d.Driver.Open(dsn)
+	if err != nil {
+		return nil, err
+	}
+	attr, err := newDBAttribute(context.Background(), d.baseName, d.Driver, rawConn, dsn, false)
+	if err != nil {
+		rawConn.Close()
+		return nil, err
+	}
+
+	conn := &driverConn{
+		Conn: rawConn,
+		attr: attr,
+	}
+	return conn, nil
+}
+
+type driverConn struct {
+	driver.Conn
+	attr *dbAttribute
+}
+
+func (conn *driverConn) Ping(ctx context.Context) error {
+	return Capture(ctx, conn.attr.dbname, func(ctx context.Context) error {
+		conn.attr.populate(ctx, "PING")
+		if p, ok := conn.Conn.(driver.Pinger); ok {
+			return p.Ping(ctx)
+		}
+		return nil
+	})
+}
+
+func (conn *driverConn) Prepare(query string) (driver.Stmt, error) {
+	panic("not supported")
+}
+
+func (conn *driverConn) PrepareContext(ctx context.Context, query string) (driver.Stmt, error) {
+	var stmt driver.Stmt
+	var err error
+	if connCtx, ok := conn.Conn.(driver.ConnPrepareContext); ok {
+		stmt, err = connCtx.PrepareContext(ctx, query)
+	} else {
+		stmt, err = conn.Conn.Prepare(query)
+		if err == nil {
+			select {
+			default:
+			case <-ctx.Done():
+				stmt.Close()
+				return nil, ctx.Err()
+			}
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+	return &driverStmt{
+		Stmt:  stmt,
+		attr:  conn.attr,
+		query: query,
+		conn:  conn,
+	}, nil
+}
+
+func (conn *driverConn) Begin() (driver.Tx, error) {
+	panic("not supported")
+}
+
+func (conn *driverConn) BeginTx(ctx context.Context, opts driver.TxOptions) (driver.Tx, error) {
+	var tx driver.Tx
+	var err error
+	if connCtx, ok := conn.Conn.(driver.ConnBeginTx); ok {
+		tx, err = connCtx.BeginTx(ctx, opts)
+	} else {
+		if opts.Isolation != driver.IsolationLevel(sql.LevelDefault) {
+			return nil, errors.New("xray: driver does not support non-default isolation level")
+		}
+		if opts.ReadOnly {
+			return nil, errors.New("xray: driver does not support read-only transactions")
+		}
+		tx, err = conn.Conn.Begin()
+		if err == nil {
+			select {
+			default:
+			case <-ctx.Done():
+				tx.Rollback()
```

Comment:
We should check for error return value.

Reply:
we can't do anything if we get an error.
so, just ignore it here.